pkg/apis/jenkins.io/v1/types_pipelinestructure.go

```diff
@@ -0,0 +1,665 @@
+	// +optional
+	Steps []PipelineStructureStep `json:"steps,omitempty" protobuf:"bytes,5,opt,name=steps"`
+	// +optional
+	Stages []*PipelineStructureStage `json:"stages,omitempty" protobuf:"bytes,6,opt,name=stages"`
+	// +optional
+	Parallel []*PipelineStructureStage `json:"parallel,omitempty" protobuf:"bytes,7,opt,name=parallel"`
+	// +optional
+	Post []PipelineStructurePost `json:"post,omitempty" protobuf:"bytes,8,opt,name=post"`
+	// +optional
+	Parent *string `json:"parent,omitempty" protobuf:"bytes,9,opt,name=parent"`
+	// +optional
+	Depth int8 `json:"depth,omitempty" protobuf:"bytes,10,opt,name=depth"`
+}
+
+// PostCondition is used to specify under what condition a post action should be executed.
+type PostCondition string
+
+// Probably extensible down the road
+const (
+	PostConditionSuccess PostCondition = "success"
+	PostConditionFailure PostCondition = "failure"
+	PostConditionAlways  PostCondition = "always"
+)
+
+var allPostConditions = []PostCondition{PostConditionAlways, PostConditionSuccess, PostConditionFailure}
+
+// PipelineStructurePost contains a PostCondition and one more actions to be executed after a pipeline or stage if the condition is met.
+type PipelineStructurePost struct {
+	// TODO: Conditional execution of something after a Task or Pipeline completes is not yet implemented
+	Condition PostCondition                 `json:"condition" protobuf:"bytes,1,opt,name=condition"`
+	Actions   []PipelineStructurePostAction `json:"actions" protobuf:"bytes,2,opt,name=actions"`
+}
+
+// PipelineStructurePostAction contains the name of a built-in post action and options to pass to that action.
+type PipelineStructurePostAction struct {
+	// TODO: Notifications are not yet supported in Build Pipeline per se.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// Also, we'll need to do some magic to do type verification during translation - i.e., this action wants a number
+	// for this option, so translate the string value for that option to a number.
+	// +optional
+	Options map[string]string `json:"options,omitempty" protobuf:"bytes,2,opt,name=options"`
+}
+
+var _ apis.Validatable = (*PipelineStructure)(nil)
+
+// TaskName translates the stage name
+func (s *PipelineStructureStage) TaskName() string {
+	return strings.ToLower(strings.NewReplacer(" ", "-").Replace(s.Name))
+}
+
+// MangleToRfc1035Label - Task/PipelineStructureStep names need to be RFC 1035/1123 compliant DNS labels, so we mangle
+// them to make them compliant. Results should match the following regex and be
+// no more than 63 characters long:
+//     [a-z]([-a-z0-9]*[a-z0-9])?
+// cf. https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+// body is assumed to have at least one ASCII letter.
+// suffix is assumed to be alphanumeric and non-empty.
+// TODO: Combine with kube.ToValidName (that function needs to handle lengths)
+func MangleToRfc1035Label(body string, suffix string) string {
```

Comment:
nit: Now that this code is in a different package, I think we can import `github.com/jenkins-x/jx/pkg/kube` without cycles and replace uses of this method with `kube.ToValidNameTruncated(foo, 63-len(suffix)-1)+"-"+suffix`.

Reply:
Will do.

Reply:
Amusingly, we're only using the suffix bit of this in tests. =)

Reply:
Yeah, I'm still thinking about how exactly we want to deal with this, especially given that Tektōn will truncate our names and add its own suffix in some cases...